### PR TITLE
When uploading new avatar/banner, delete old one

### PR DESCRIPTION
**Pull request by Nutomic**
We currently keep old avatars, banners etc around forever. It makes sense to delete them when a new one is uploaded. For now I implemented it only for user avatar change, but will do the same for user banner, community icon etc if it looks good.

Depends on https://github.com/LemmyNet/lemmy/pull/4509

### CODE REVIEW

File: crates/db_schema/src/impls/images.rs [hunk 1]
```diff
@@ -80,6 +82,17 @@ impl LocalImage {
       .execute(conn)
       .await
   }
+
+  pub async fn delete_by_url(pool: &mut DbPool<'_>, url: &DbUrl) -> Result<Self, Error> {
+    let conn = &mut get_conn(pool).await?;
+    static IMAGE_REGEX: Lazy<Regex> =
+      Lazy::new(|| Regex::new(r"^.*/pictrs/image/([a-z0-9-]+\.[a-z]+)$").expect("compile regex"));
```
Comment (Nutomic) [inline crates/db_schema/src/impls/images.rs:89]: This isnt very elegant but I dont see any better way unless we separately store the alias for each avatar image.

File: crates/api/src/local_user/save_settings.rs [hunk 2]
```diff
@@ -40,6 +42,17 @@ pub async fn save_user_settings(
...
+    // Ignore errors because image may be stored externally.
+    if let Some(avatar) = &local_user_view.person.avatar {
+      let image = LocalImage::delete_by_url(&mut context.pool(), &avatar)
+        .await
+        .ok();
+      if let Some(image) = image {
+        delete_image_from_pictrs(&image.pictrs_alias, &image.pictrs_delete_token, &context).await?;
+      }
+    }
+  }
```
Comment (Nutomic) [inline crates/api/src/local_user/save_settings.rs:55]: This will go into a helper function.

File: crates/db_schema/src/impls/images.rs (hunk 1 above)
Comment (dessalines) [reply crates/db_schema/src/impls/images.rs:89]: Regex is overkill here IMO. 

A pictrs avatar will always be at something like: `https://lemmy.world/pictrs/image/78b09748-85b9-4678-8375-096b57595903.jpeg` . So just split by `/`, pop the last one, and remove the extension to get the alias.

I'd also recommend this function not running the diesel::delete, but just extracting the alias, and running `Self::delete_by_alias`

File: crates/api/src/local_user/save_settings.rs (hunk 2 above)
Comment (dessalines) [reply crates/api/src/local_user/save_settings.rs:55]: Thanks, I was just thinking of commenting abt that. Makes this a lot cleaner, even if it is only used once.

**Comment by Nutomic** [comment]
Updated, it now also deletes old site and community icons/banners when a new one is uploaded.

File: crates/db_schema/src/impls/images.rs (hunk 1 above)
Comment (Nutomic) [reply crates/db_schema/src/impls/images.rs]: Makes sense, fixed.

File: crates/db_schema/src/impls/images.rs [hunk 3]
```diff
@@ -74,12 +74,17 @@ impl LocalImage {
...
+  pub async fn delete_by_alias(pool: &mut DbPool<'_>, alias: &str) -> Result<Self, Error> {
     let conn = &mut get_conn(pool).await?;
     diesel::delete(local_image::table.filter(local_image::pictrs_alias.eq(alias)))
-      .execute(conn)
+      .get_result(conn)
       .await
   }
+
+  pub async fn delete_by_url(pool: &mut DbPool<'_>, url: &DbUrl) -> Result<Self, Error> {
+    let alias = url.as_str().split('/').last().ok_or(NotFound)?;
```
Comment (dessalines) [inline crates/db_schema/src/impls/images.rs:85]: Does the alias include the file extension? I spose your tests would fail otherwise.

**Review by dessalines — APPROVED** [review]
Looks good, feel free to merge after fixing that clippy issue.

File: crates/db_schema/src/impls/images.rs (hunk 3 above)
Comment (Nutomic) [reply crates/db_schema/src/impls/images.rs:85]: Yes with extension.

**Comment by dessalines** [comment]
This maybe was wrongly closed cause that branch got merged and deleted. @Nutomic you might need to re-open this up, but pointing at `main`

**Comment by SleeplessOne1917** [comment]
Shit, This was not what I meant to close.

**Comment by Nutomic** [comment]
No worries it was only deleted because the delete_images branch was deleted. I have no clue why the PR pointed at that and not main.

**Pull request closed without merging**